csharp/src/Google.Protobuf/Reflection/FieldDescriptor.cs

```diff
@@ -196,6 +196,11 @@ private static FieldType GetFieldTypeFromProtoType(FieldDescriptorProto.Types.Ty
         /// </summary>
         public bool IsRequired => Proto.Label == FieldDescriptorProto.Types.Label.Required;
 
+        /// <summary>
+        /// Returns <c>true</c> if this field is a proto3 optional field; <c>false</c> otherwise.
+        /// </summary>
+        public bool IsProto3Optional => Proto.Proto3Optional;
```

Comment:
I don't want to expose this publicly. I think we paint ourselves into a corner if code is written to think about the difference between proto2 and proto3.

Could you follow the pattern of the C++ API? C++ is hiding this accessor, and instead exposing a more semantically meaningful accessor: `has_presence()`.